dlib/internal/integer_utils.cpp

```diff
@@ -0,0 +1,42 @@
+// Copyright (C) 2019 Paul Dreik (github@pauldreik.se)
+// License: Boost Software License   See LICENSE.txt for the full license.
+#include "integer_utils.h"
+
+#include <stdexcept>
+#include <cstdint>
+#include <limits>
+
+namespace dlib
+{
+namespace detail {
+long safe_multiply(long a, long b) {
+    long ret;
+    bool overflow=false;
+#ifdef DLIB_HAS_BUILTIN_MUL_OVERFLOW
```

Comment:
This should be done with a `#ifdef` check for the compiler version, rather than using a macro and cmake rule.  This is to simplify the build system.  Not everyone uses cmake.  You would also have to update the cmake install scripts if we really went with `DLIB_HAS_BUILTIN_MUL_OVERFLOW` so that `config.h` is properly handled as well.  But this sort of thing could be done more simply using `#ifdef such and such gcc or clang version` style